FEAT: add support for AGIPD500K2G detector layout

tmichela opened the pull request:
![Screenshot from 2020-09-22 17-04-18](https://user-images.githubusercontent.com/32831491/94005146-2673ba00-fd9e-11ea-9553-4758e2fdcf38.png)

![Screenshot from 2020-09-22 17-04-31](https://user-images.githubusercontent.com/32831491/93901011-5ebfbd80-fcf6-11ea-8278-b42767512eaa.png)

`module 0` is from recorded data (others are random), you can see the first row of ASICS is rotated by 180 degree

takluyver commented:
Looks promising. It would be good to have at least a couple of basic tests with this.

Should we use `ax.set_aspect(1)` in inspect() to make the shape clearer? The first plot is stretching the panels vertically quite a bit.

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +854,170 @@ def to_distortion_array(self, allow_negative_xy=False):
+        This produces an idealised geometry, assuming all modules are perfectly
+        flat, aligned and equally spaced within their quadrant.
+
+        The center position is given in pixel units, referring to the first
+        pixel of the first module, corresponding to data channels 0.
+
+        The origin of the coordinates is the bottom-right corner of the
+        detector. Coordinates increase upwards and to the left (looking along
+        the beam).
```
This is a bit confusing. It appears to be saying that you give the location of M0 with respect to the bottom beam-right corner (by M6), but that's not the case.

We must be either giving the location of M0 with respect to the centre (the point where a ray from the sample would hit perpendicular to the detector plane) or vice versa - the position of the centre with respect to some fixed point on the detector hardware. I think the implementation is the former, but calling this `from_center` suggests the latter.

tmichela replied:
Thanks, I updated the text, hope it's clearer.

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +854,170 @@ def to_distortion_array(self, allow_negative_xy=False):
+        # Label modules and tiles
+        for ch, module in enumerate(self.modules):
+            cx, cy, _ = module[4].centre() * scale
+            ax.text(cx, cy, f'Mod{ch}', fontweight='bold',
```
Is it the convention that we start with 0 for this one, not from 1 like the Q1M1 numbers for AGIPD-1M?

tmichela replied:
Good question. I've identified the modules with the channel number for now, I'll ask DET.

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +854,170 @@ def to_distortion_array(self, allow_negative_xy=False):
+        ax.set_title(f'AGIPD-500K2G detector geometry ({self.filename})')
+        return ax
+
+    def _get_dimensions(self):
```
I don't think we actually need this unless we define a method to assemble data by interpolation - we did that for AGIPD-1M, but it's horrendously slow (maybe there's a better way to do it that I didn't find).

tmichela replied:
I added the `position_modules_interpolate` method

tmichela commented:
> Looks promising. It would be good to have at least a couple of basic tests with this.
> 
> Should we use `ax.set_aspect(1)` in inspect() to make the shape clearer? The first plot is stretching the panels vertically quite a bit.

Thanks, I fixed the aspect ration and added unit test.

tmichela commented:
I removed the `position_modules_interpolate`, it wasn't working properly. I'll add it if it's required at some point, but on't think we need it at all...

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +856,159 @@ def to_distortion_array(self, allow_negative_xy=False):
+
+    def position_modules_fast(self, *args, **kwargs):
+        data, _ = super().position_modules_fast(*args, **kwargs)
+        return data, (0, 0)  # center is the origin
```
I don't think this is quite right. The aim of returning the centre here is to say where the origin of the detector layout you've given EXtra-geom falls *in the output array*. It doesn't make sense to hard-code it.

tmichela replied:
I take now the center of the output array size

takluyver replied:
I still think this is not quite right. I guess 'center' is a misnomer - the idea is to return the location corresponding to the origin in the geometry object, i.e. the PONI. If the caller wants `(height / 2, width / 2)`, they can easily get that from the array shape.

tmichela replied:
Oh, yeah. I agree. I misunderstood what you meant, will change taht

tmichela replied:
Hum, thinking about it

>  the idea is to return the location corresponding to the origin in the geometry object, i.e. the PONI.

In that case, it should always be `(0, 0)`, right. Or I did not get it... :thinking:

takluyver replied:
Sorry, should have been more specific. It's the origin from the geometry object, in the coordinates of the output array. So, for this class, I'd expect the `from_origin(origin=` parameter - in pixel units from the low-x, low-y corner - to be returned as the centre position in the array as well (but reversed, because array indexing goes (y, x)).

The idea is to tell you where the theoretical origin/PONI is in the assembled image.

tmichela replied:
Aha, thanks. Hope I got that right now :D

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +856,159 @@ def to_distortion_array(self, allow_negative_xy=False):
+        This produces an idealised geometry, assuming all modules are perfectly
+        flat, aligned and equally spaced within their quadrant.
+
+        The center position is given in pixel units, referring to the first
+        pixel of the first module, corresponding to data channels 0.
+
+        The origin of the coordinates is the bottom-right corner of the
+        first module of the detector (channel 0). Coordinates increase upwards
+        and to the left (looking along the beam).
```
I still think this is a muddle. The whole notion of the 'center' is kind of confusing here. I'll try to explain:

For the most part, what we're interested in is where all the bits of the detector are relative to one another, so we can assemble a complete image. In the big quadrant-based detectors, we use the centre (where the beam goes through a hole) as the origin, and give the quadrant positions relative to that. This detector doesn't have an obvious central feature, and for assembling an image, it doesn't matter where we put the origin.

But for a second level of analysis, the origin does matter. E.g. if you're doing azimuthal integration, you need a point to integrate around. We've referred to the point where the beam passes through the detector, which works for a detector perpendicular to the beam, but more generally, we want the origin to be where a ray from the sample is perpendicular to the detector plane. PyFAI calls this the PONI, or Point Of Normal Incidence.

I'm not exactly sure how to handle this. Maybe we should have a `AGIPD_500K2GGeometry.default()` class method which constructs it with the origin in the centre of the 8 modules (as an arbitrary choice)?  Do we know anything about how it will be positioned for use?

tmichela replied:
I jsut asked. It will be sitting on a bench at the end of the chamber, perpendicular to the beam direction (similar to SPB). Beam will pass probably above or below the detector. Also learned that the 500K agipd is temporary they will install a full AGIPD at some point.

Do you mean to have 2 methods: 1 `default` from the center (or corner, ..) + 1 from the PONI (how we should call that)?

takluyver replied:
I think that might be the best thing. `default()` with no required arguments, so you can assemble an image without worrying about where the detector is, and something like `from_center` or `from_poni`. I can't decide if 'center' is meaningful enough, or 'PONI' is too much jargon. Maybe it's better to use a word people don't know than one which they might guess an incorrect meaning for.

I also wonder if this second method should invert the numbers, i.e. you give the position of the PONI relative to some part of the detector, rather than giving the position of the sensor relative to the PONI.

tmichela replied:
So, I added a `default` method. with the reference (0, 0) in the bottom-right corner (so pixel indices are always positive), I can change that to the centrer or an other corner if you prefer...
I also changed `from_center` to `from_origin`, is that an OK name for you? (also inverted the numbers: origin position is relative to the detector corner)

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +856,191 @@ def to_distortion_array(self, allow_negative_xy=False):
+        """Generate an AGIPD-500K2G geometry from origin position.
+
+        This produces an idealised geometry, assuming all modules are perfectly
+        flat, aligned and equally spaced within their quadrant.
```
```suggestion
        flat, aligned and equally spaced within the detector.
```

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +856,191 @@ def to_distortion_array(self, allow_negative_xy=False):
+        This produces an idealised geometry, assuming all modules are perfectly
+        flat, aligned and equally spaced within their quadrant.
+
+        The origin of the coordinates is relative to the bottom-right corner of
+        the detector. Coordinates increase upwards and to the left (looking
+        along the beam).
```
```suggestion
        The default origin is in the bottom-right corner of the detector.
        Coordinates increase upwards and to the left (looking along the beam).
        The origin parameter shifts the origin by the given (x, y) offset.
        For some kinds of analysis, the origin should be at the point of normal
        incidence, where a ray from the sample is perpendicular to the detector
        plane.
```

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +856,191 @@ def to_distortion_array(self, allow_negative_xy=False):
+    n_tiles_per_module = 8
+
+    @classmethod
+    def default(cls, asic_gap=2, panel_gap=(16, 30), unit=pixel_size):
```
Looking at this now, maybe it's clearer to have just `from_origin`, with a default so the `origin` parameter is optional. What do you think?

tmichela replied:
Yep, that's good

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +856,191 @@ def to_distortion_array(self, allow_negative_xy=False):
+    def position_modules_fast(self, *args, **kwargs):
+        return super().position_modules_fast(*args, **kwargs)
+
+    def _snapped(self):
+        """Snap geometry to a 2D pixel grid
+
+        This returns a new geometry object. The 'snapped' geometry is
+        less accurate, but can assemble data into a 2D array more efficiently,
+        because it doesn't do any interpolation.
+        """
+        if self._snapped_cache is None:
+            snapped = super()._snapped()
+            snapped.centre = np.array(snapped.size_yx) // 2
+        return self._snapped_cache
```
I think don't override this, so `centre` is calculated the same way as for the other detectors. Maybe we should rename it to something like `poni` or `geom_origin`, but I don't think it fundamentally needs to be different for this detector.

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +856,171 @@ def to_distortion_array(self, allow_negative_xy=False):
+        super().write_crystfel_geom(*args, nquads=1, **kwargs)
+
+    def position_modules_fast(self, *args, **kwargs):
+        data, _ = super().position_modules_fast(*args, **kwargs)
+        origin = data.shape - self.origin[::-1]
```
I think this should be unnecessary. What we're trying to achieve is the same logical operation that the other detectors already do to get the origin ('centre') location for this method.

Have you tried just using the base class' `position_modules_fast()` method? Does that not behave as expected? I can't see that it should need overriding.

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +856,171 @@ def to_distortion_array(self, allow_negative_xy=False):
+        flat, aligned and equally spaced within the detector.
+
+        The default origin (0, 0) of the coordinates is the bottom-right corner
+        of the detector. If an other coordinate is given as the origin, it is
```
```suggestion
        of the detector. If another coordinate is given as the origin, it is
```

takluyver commented on extra_geom/detectors.py:
```diff
@@ -854,6 +856,171 @@ def to_distortion_array(self, allow_negative_xy=False):
+        of the detector. If an other coordinate is given as the origin, it is
+        relative to the bottom-right corner. Coordinates increase upwards and
+        to the left (looking along the beam).
+
+        The origin of the coordinates is relative to the bottom-right corner of
+        the detector. Coordinates increase upwards and to the left (looking
+        along the beam).
```
```suggestion
```

takluyver commented:
Thanks!

tmichela commented:
\o/

thanks for the review :)
